CIPs/cip-120.md

```diff
@@ -0,0 +1,109 @@
+
+### Example
+
+If a stream has the following genesis header,
+
+```javascript
+{
+  header: {
+    sep: 'model',
+    model: <bytes-k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi>
+  },
+  ...
+}
+```
+
+it would use the following pubsub topic on main net:
+
+```markdown
+/ceramic/mainnet/model/k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi
+```
+
+### Streams without `sep`
+
+For backwards compatibility if a stream doesn't have a `sep` field or no value for the field specified by `sep`,  will use the following pubsub topic:
+
+```markdown
+/ceramic/<network-name>/
+```
+
+Which is the pubsub topic currently used by Ceramic.
+
+### Peer discovery
+
+While gossipsub used by Ceramic enables ambient discovery of peers, an initial set of peers already in a topic is required to recieve messages. It is therefore recommended that peers announce to the libp2p DHT that they are subscribed to a given pubsub topic. Peers can thus also discover other peers by looking up the DHT records for the given topic. The key that should be used for DHT providing and lookup MUST be `sha2-256(separator-key || separator-value)`.
+
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+By separating tip gossip into multiple pubsub topics, each node only needs to hear about a smaller subset of messages. Each node will subscribe only to the topics that are relevant for the streams it currently cares about.
+
+
+## Backwards Compatibility
+<!--All CIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The CIP must explain how the author proposes to deal with these incompatibilities. CIP submissions without a sufficient backwards compatibility section may be rejected outright.-->
+
+There are two approaches for how to transition to this new approach for pubsub topics.
+
+### Duplicate messages
+
+For a transition period new nodes will publish updates in both the old main topic as well as the new topic. This ensures backwards compatibility in the system at the cost of some amount of time with, at worst, the amount of messages.
+
+### Force an upgrade (recommended)
+
+Almost like a hard fork we can pick a particular blockheight (since we track the eth blockchain), or blocktimestamp. Once this time has passed all nodes would switch over to the new pubsub topics (with maybe a short phase out period). The main drawback of this approach is that the community would need to coordinate to upgrade their software in time. There would be no backwards compatibility for nodes that didn’t make the upgrade.
```

Comment:
if we do this before composedb is supported on mainnet, and only do it for Model and MID streams, do we even need to do anything for backwards compatibility at all?

Reply:
I guess this also raises the question - do we want to do anything to TileDocument and Caip10Link streams?  Give them their own dedicated topic?  I guess we can't change the genesis contents without breaking determinism.  Makes me wonder if we should also encode the streamtype into the topic name, so streams with different streamtypes are always in a different pubsub topic.

Reply:
I don't think we will introduce multiple pubsub topics before mainnet. Introducing the `sep` field I think we should do before mainnet, but this section talks more about the pubsub specific changes.

For Tile and Caip10Link I think we should not change anything, except that future Tiles can introduce the `sep`  field.

Reply:
we can't add a `sep` field to tiles without breaking determinism.  What do you think about putting the streamtype into the topic name?  That would ensure that no Streams are using the top-level Ceramic topic, and let us keep that topic reserved for system-wide messages (like the keepalive messages used to share node versions, for example)

Reply:
I guess it does beg the question, are streamtypes something that the base event streaming layer should even be aware of?  At first glance it feels like that should be a higher-level concern, but it also kind of feels like it will be impossible for the base layer to avoid needing to know at least something about the streamtypes.  Especially as the type ID numbers are a global namespace and we'll need some registry of streamtypes to prevent collisions.

Reply:
> we can't add a sep field to tiles without breaking determinism

Not for existing apps, but for all new ones this wouldn't be a problem.

>  That would ensure that no Streams are using the top-level Ceramic topic, and let us keep that topic reserved for system-wide messages

If a node is only subscribed to one separator (e.g. model) I would assume they don't need to subscribe to a "top-level" topic, only the specific topic which they care about. It's probably a good idea to run keep alive messages on all subscribed topics for this reason. 

> but it also kind of feels like it will be impossible for the base layer to avoid needing to know at least something about the streamtypes

Would be curious to hear more about what you are thinking about this. 
Overall right now I don't see a strong reason to add streamtypes to the topic. Seems like a separate concern to me.

>  Especially as the type ID numbers are a global namespace and we'll need some registry of streamtypes to prevent collisions.

StreamType numbers are registered in the [StreamType table](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-59/CIP-59.md)

Reply:
> Not for existing apps, but for all new ones this wouldn't be a problem.

How would that work?  Each app would have to opt-in in their app code to using the new pubsub channel?  If we just changed the behavior in the TileDocument package it would affect all apps, old and new.

> If a node is only subscribed to one separator (e.g. model) I would assume they don't need to subscribe to a "top-level" topic

Maybe, maybe not.  I could definitely imagine many potential uses for a global topic that all Ceramic nodes subscribe to.  But I guess if that did become a need we could always create a new dedicated topic it for it rather than using the current topic.  We could leave the current topic to just be used by TileDocument and Caip10Link.  So maybe it's fine to leave streamtype out of the topic name.

> StreamType numbers are registered in the [StreamType table](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-59/CIP-59.md?rgh-link-date=2023-01-03T21%3A22%3A14Z)

That is a highly centralized approach.  I had assumed that at some point we'd want to move away from that to something more decentralized, and that whatever we do there might at least conceivably require Ceramic base layer support.  But I could also imagine solutions that don't require base layer support, so I think I agree that it makes sense to try to leave StreamTypes out of the base layer for now.

Reply:
only if they built their genesis commit manually.  The standard TileDocument APIs don't allow this: https://github.com/ceramicnetwork/js-ceramic/blob/b443590b6e346d5808e72d96fa868ec62114bff0/packages/stream-tile/src/tile-document.ts#L238

Reply:
Ok, this would be an easy code change though!

Reply:
Yep of course.  That's all I was trying to confirm with my original statement: app devs would have to manually opt-in to the new behavior.  We couldn't make the change for them.